### PR TITLE
gadgets: add gadget to trace mutex locks/unlocks

### PR DESCRIPTION
**Pull request by Snehil-Shah**
# Gadget to trace mutex locks/unlocks

Adds a basic gadget that traces mutex acquisitions and releases. This gadget will be extended to detect potential deadlocks.

## How to use

1. Run the gadget using `ig`.
2. Run a simple script that uses the `pthread_mutex_lock` and `pthread_mutex_unlock` methods, in a container.
3. Observe trace.

## Testing done

Followed the above steps with the following `test` [script](https://github.com/iovisor/bcc/blob/master/tools/deadlock_example.txt#L187).

**Output**:

image

### CODE REVIEW

**Review by alban — COMMENTED** [review]
Thanks, the code looks good to me.

The CI reveals issues to fix:
- run `make clang-format`
- add artifacthub-pkg.yml
- add README.md and README.mdx
- add symlink in docs/gadgets/

**Comment by Snehil-Shah** [comment]
The links I have added don't exist yet. Should I leave them empty? Seems like it's one of the reasons CI is failing.

**Comment by alban** [comment]
> The links I have added don't exist yet. Should I leave them empty? Seems like it's one of the reasons CI is failing.

Yes, the "Documentation checks" step in the CI is fine to ignore. I had a look and the only dead link will be resolved after it is merged.

> ERROR: 1 dead links found!
> [✖] https://inspektor-gadget.io/docs/latest/gadgets/deadlock → Status: 404

But the "Gadget guides check" step in the CI should be fixed. You should be able to run `make -C gadgets/ readmes` after installing [gomplate](https://docs.gomplate.ca/installing/) and check the result with `git grep sidebar_position -- gadgets/`.

The other CI failures seem temporary. I can restart the CI after the "Gadget guides check" step above is fixed.

File: gadgets/deadlock/README.mdx
Comment (alban) [inline gadgets/deadlock/README.mdx:39]:
```suggestion
For this example, we use a [test C++ program](https://github.com/iovisor/bcc/blob/master/tools/deadlock_example.txt#L187) (from BCC) with lock inversions that can cause a potential deadlock.
```

File: gadgets/deadlock/README.mdx
Comment (alban) [inline gadgets/deadlock/README.mdx:41]:
```suggestion
The deadlock gadget can trace all mutex lock/unlock events in the following way:
```

**Review by alban — APPROVED** [review]
2 phrasing suggestions. After that, LGTM, thanks!

I'll wait for the CI to complete before merging.

**Comment by alban** [comment]
Did you forget to git-add all the `gadgets/*/README.mdx` after running `make -C gadgets readmes`?

**Comment by Snehil-Shah** [comment]
@alban Yeah, did the needfull👍

**Comment by alban** [comment]
The CI still fails. From the CI:
>  HEAD is now at 8b56975 Merge ddf8e67c4be95dd4d397a2eaeb65bf999b1cb9e3 into 3ecf4130667817240d29db38860776643048d5e9

It seems the CI didn't test your branch directly but the result of merging main into your branch.

And 4795e94479f14ce571b19f6131c633345e93aff8 (#3452) was merged since your prepared your branch. So the gadget numbering changed on the main branch.

Could you rebase your branch on origin/main and update the readmes again? Thanks.

**Comment by Snehil-Shah** [comment]
@alban Done. Should be fine now.

**Pull request merged**
